keps/0008-previous-values-in-update-plans-and-conditional-phases.md

```diff
@@ -0,0 +1,35 @@
+---
+kep-number: 8
+title: Previous values in update plans and conditional phases
+authors:
+  - "@djannot"
+owners:
+  - TBD
+editor: TBD
+creation-date: 2019-04-01
+status: provisional
+---
+
+# Previous values in update plans and conditional phases
+
+## Table of Contents
+
+  * [Table of Contents](#table-of-contents)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+
+Created by [gh-md-toc](https://github.com/ekalinin/github-markdown-toc)
+
+## Summary
+
+In order to develop the logic behind an `update` `Plan`, you generally need to know what were the previous values.
+
+For example, if the number of replicas of a `StatefulSet` is updated, you ofter need to add a new `Phase` to reconfigure your cluster.
+
+But currently, you don't know what has changed. You don't know if the number of replicas has changed and if it has changed you don't know what was the previous value.
+
+We would need a way to access the previous values, but also probably to be able to define `Phases` that need to be run based on conditions.
```

Comment:
IMO KUDO should incentivize and encourage building robust Packages, including not encouraging anti-patterns in K8s. In the example of a Redis operator and the case when the number of replicas is updated I think providing the [option to trigger custom plans](https://github.com/kudobuilder/kudo/blob/master/config/samples/mysql.yaml#L22-L27) enables App Developer and Operator to build such a robust Package. I can't think on the top of my head right now of a case that couldn't be solved right now, but if you have one maybe you can give an example.

Right now this type of operations are handled in the `planexecution_controller.go` with this switch logic https://github.com/kudobuilder/kudo/blob/431aa5f2cc6c6d2678aa8a8ff00cd108a19676c6/pkg/controller/planexecution/planexecution_controller.go#L492-L535

The sticky knowledge of what to do in such changes would live around here but considering all the edge-cases it can become quick messy.

Reply:
@fabianbaier the problem with the option to trigger custom plans is that changing several environment variables will trigger several plans or several time the same plan. There will be cases when several environment variables are going together (like Kerberos fields for example)

Reply:
> @fabianbaier the problem with the option to trigger custom plans is that changing several environment variables will trigger several plans or several time the same plan. There will be cases when several environment variables are going together (like Kerberos fields for example)

It would trigger multiple times the same plan, that's right. But how would having access to previous values solve that? What we would need is some sort of grouping or group triggers, a collector that aggregates those environment variables that fall under the same plan before it is actually triggered.

Reply:
If you have a single plan launched when an update is performed (whatever is updated, one or several values), the logic can be added in a unique plan.
But I also see the value of having different plans triggered for each different value. so not sure about what is the best solution. I think that implementing KEP 006 will help identifying the best approach based on a real use case.